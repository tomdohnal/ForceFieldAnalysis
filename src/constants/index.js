export const MIN_STRENGTH = 1;
export const MAX_STRENGTH = 8;
