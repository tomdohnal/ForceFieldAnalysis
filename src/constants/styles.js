export const HEADER_HEIGHT = '60px';
export const FOOTER_HEIGHT = '80px';

export const COLORS = {
  WHITE: '#fff',
  BLACK: '#000',
  BLUE: '#268bd2',
  RED: '#dc2f2f',
  GREEN: '#859900',
};
