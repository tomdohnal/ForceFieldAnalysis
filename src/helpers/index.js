// @flow
type Action = {
  type: string,
  payload?: any,
}

export type Reducer<S, A: Action> = (S, A) => S;

export type ReduxHandlers<S, A: Action> = {
  [key: string]: Reducer<S, A>,
}

export const createReducer =
  <S, A: Action>(initialState: S, handlers: ReduxHandlers<S, A>) =>
    (state: S = initialState, action: A) => (
      Object.prototype.hasOwnProperty.call(handlers, action.type) ?
        handlers[action.type](state, action) : state
    );

export const moveCaretAtTheEnd = (e: SyntheticEvent<HTMLInputElement>) => {
  if (e.target instanceof HTMLInputElement) {
    const { value } = e.target;
    e.target.value = '';
    e.target.value = value;
  }
};
