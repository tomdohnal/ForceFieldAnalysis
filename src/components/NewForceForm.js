// @flow
import React, { Component } from 'react';
import { Form, Input, Button, Label } from 'semantic-ui-react';

import { MIN_STRENGTH, MAX_STRENGTH } from '../constants/index';
import { COLORS } from '../constants/styles';
import { Text } from './common/index';

type State = {
  newForceName: string,
  newForceError: boolean,
  newForceErrorText: string,
  newForceStrength: number,
};

type Props = {
  submitButtonText: string,
  submitButtonColor: string,
  onFormSubmit: (name: string, strength: number) => void,
};

type InputData = {
  value: string,
}

class NewForceForm extends Component<Props, State> {
  defaultState = {
    newForceName: '',
    newForceError: false,
    newForceErrorText: '',
    newForceStrength: 1,
  };

  state = this.defaultState;

  onNewForceInputChange = (e: SyntheticEvent<HTMLInputElement>, { value }: InputData) => {
    this.setState({
      newForceName: value,
      newForceError: false,
      newForceErrorText: '',
    });
  };

  onIncreaseForceButtonClick = () => {
    if (this.state.newForceStrength !== MAX_STRENGTH) {
      this.setState(prevState => ({
        newForceStrength: prevState.newForceStrength + 1,
      }));
    }
  };

  onDecreaseForceButtonClick = () => {
    if (this.state.newForceStrength !== MIN_STRENGTH) {
      this.setState(prevState => ({
        newForceStrength: prevState.newForceStrength - 1,
      }));
    }
  };

  onNewForceButtonClick = () => {
    if (!this.state.newForceName) {
      this.setState({
        newForceError: true,
        newForceErrorText: 'Enter the description of the force.',
      });
    } else {
      this.props.onFormSubmit(this.state.newForceName, this.state.newForceStrength);

      this.setState(this.defaultState);
    }
  };

  render() {
    const { submitButtonText, submitButtonColor } = this.props;

    return (
      <Form>
        <Form.Field>
          <label>Force name:</label>
          <Input
            onChange={this.onNewForceInputChange}
            value={this.state.newForceName}
            error={this.state.newForceError}
          />
          {this.state.newForceError ? <Label basic color="red" pointing>{this.state.newForceErrorText}</Label> : ''}
        </Form.Field>
        <Form.Group inline>
          <label style={{ whiteSpace: 'nowrap' }}>Force strength:</label>
          <Button
            disabled={this.state.newForceStrength === MAX_STRENGTH}
            basic
            onClick={this.onIncreaseForceButtonClick}
            content="Increase"
            icon="plus"
            labelPosition="left"
            as="a"
          />
          <Text fontSize="24px" padding="0 12px">{this.state.newForceStrength}</Text>
          <Button
            disabled={this.state.newForceStrength === MIN_STRENGTH}
            basic
            onClick={this.onDecreaseForceButtonClick}
            content="Decrease"
            icon="minus"
            labelPosition="right"
            as="a"
          />
        </Form.Group>
        <Form.Field>
          <Button
            fluid
            onClick={this.onNewForceButtonClick}
            style={{ color: COLORS.WHITE, background: submitButtonColor }}
          >
            {submitButtonText}
          </Button>
        </Form.Field>
      </Form>
    );
  }
}

export default NewForceForm;
