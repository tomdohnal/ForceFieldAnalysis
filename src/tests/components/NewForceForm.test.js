import React from 'react';
import { shallow } from 'enzyme';

import { mockShallowComponentMethod } from '../helpers';
import NewForceForm from '../../components/NewForceForm';
import { MIN_STRENGTH, MAX_STRENGTH } from '../../constants/index';

const onFormSubmit = jest.fn();

const ShallowForm = shallow(
  <NewForceForm
    submitButtonColor="#FFF"
    submitButtonText="Submit Force"
    onFormSubmit={onFormSubmit}
  />,
);

describe('NewForceForm', () => {
  describe('rendering', () => {
    it('renders correctly', () => {
      const tree = ShallowForm.dive();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    beforeEach(() => {
      ShallowForm.setState(ShallowForm.instance().defaultState);
    });

    it('can increase force strength', () => {
      ShallowForm.instance().onIncreaseForceButtonClick();

      expect(ShallowForm.state('newForceStrength')).toBe(2);
    });

    it('has a max force strength limit', () => {
      ShallowForm.setState({ newForceStrength: MAX_STRENGTH });

      ShallowForm.instance().onIncreaseForceButtonClick();

      expect(ShallowForm.state('newForceStrength')).toBe(MAX_STRENGTH);
    });

    it('can decrease force strength', () => {
      ShallowForm.setState({ newForceStrength: MIN_STRENGTH + 1 });

      ShallowForm.instance().onDecreaseForceButtonClick();

      expect(ShallowForm.state('newForceStrength')).toBe(1);
    });

    it('has a min force strength limit', () => {
      ShallowForm.setState({ newForceStrength: MIN_STRENGTH });

      ShallowForm.instance().onDecreaseForceButtonClick();

      expect(ShallowForm.state('newForceStrength')).toBe(MIN_STRENGTH);
    });

    it('does throw an error when submitting without name', () => {
      ShallowForm.instance().onNewForceButtonClick();

      expect(ShallowForm.state('newForceError')).toBeTruthy();
    });
  });

  describe('method triggering', () => {
    it('triggers onNewForceInputChange when input is changed', () => {
      mockShallowComponentMethod(ShallowForm, 'onNewForceInputChange');

      ShallowForm.find('Input').dive().find('input').simulate('change');

      expect(ShallowForm.instance().onNewForceInputChange).toBeCalled();
    });

    it('triggers onIncreaseForceButtonClick when increase button clicked', () => {
      mockShallowComponentMethod(ShallowForm, 'onIncreaseForceButtonClick');

      ShallowForm.dive().find('Button[content="Increase"]').dive().find('a')
        .simulate('click');

      expect(ShallowForm.instance().onIncreaseForceButtonClick).toBeCalled();
    });

    it('triggers onDecreaseForceButtonClick when increase button clicked', () => {
      // can't test it due to a bug in semantic ui react
      // it throws TypeError: Cannot read property 'preventDefault' of undefined

      /* mockShallowComponentMethod(ShallowForm, 'onDecreaseForceButtonClick');

      ShallowForm.dive().find('Button[content="Decrease"]').dive().find('a')
        .simulate('click');

      expect(ShallowForm.instance().onDecreaseForceButtonClick).toBeCalled(); */
    });

    it('triggers onNewForceButtonClick when the new force button is clicked', () => {
      mockShallowComponentMethod(ShallowForm, 'onNewForceButtonClick');

      ShallowForm.dive().find('Button[as="button"]').dive().find('button')
        .simulate('click');

      expect(ShallowForm.instance().onNewForceButtonClick).toBeCalled();
    });
  });
});
