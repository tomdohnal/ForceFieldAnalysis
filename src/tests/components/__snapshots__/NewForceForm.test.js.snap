// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewForceForm rendering renders correctly 1`] = `
<form
  className="ui form"
  onSubmit={[Function]}
>
  <FormField>
    <label>
      Force name:
    </label>
    <Input
      error={false}
      onChange={[Function]}
      type="text"
      value=""
    />
  </FormField>
  <FormGroup
    inline={true}
  >
    <label
      style={
        Object {
          "whiteSpace": "nowrap",
        }
      }
    >
      Force strength:
    </label>
    <Button
      as="a"
      basic={true}
      content="Increase"
      disabled={false}
      icon="plus"
      labelPosition="left"
      onClick={[Function]}
    />
    <text
      fontSize="24px"
      padding="0 12px"
    >
      1
    </text>
    <Button
      as="a"
      basic={true}
      content="Decrease"
      disabled={true}
      icon="minus"
      labelPosition="right"
      onClick={[Function]}
    />
  </FormGroup>
  <FormField>
    <Button
      as="button"
      fluid={true}
      onClick={[Function]}
      style={
        Object {
          "background": "#FFF",
          "color": "#fff",
        }
      }
    >
      Submit Force
    </Button>
  </FormField>
</form>
`;
